### ISSUE TITLE
[Questions] About LightGBMTunerCV

### ISSUE THREAD

**Issue by rohan-gt** [optuna/optuna#1769]
1. What is the default value of n_estimators and learning rate that LightGBMTunerCV uses? Are these tuned?
2. Are the `num_boost_rounds` and `early_stopping_round` parameters for LightGBM or the tuner?
3. Is it possible to feed in additional parameters to tune and custom ranges for existing parameters?
4. Is early stopping implemented for the tuner? If so what's the early stopping logic?

**Comment by toshihikoyanase** [comment]
I'm sorry for my late response.

> 1. What is the default value of n_estimators and learning rate that LightGBMTunerCV uses? Are these tuned?

In my understanding, the default value of `n_estimators` is [`1000`](https://github.com/optuna/optuna/blob/ed661d57d450169cf45113d8463d3a65c56c19fe/optuna/integration/_lightgbm_tuner/optimize.py#L332), and the default `learning rate` is [`0.1`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#learning_rate).  These values are not the target of parameter tuning because they are not in the following list:

https://github.com/optuna/optuna/blob/ed661d57d450169cf45113d8463d3a65c56c19fe/optuna/integration/_lightgbm_tuner/optimize.py#L188-L196
 
> 2. Are the num_boost_rounds and early_stopping_round parameters for LightGBM or the tuner?

I think they are LightGBM parameters. When the tuner receives them, it directly passes them to the `LightGBM`.

> 3. Is it possible to feed in additional parameters to tune and custom ranges for existing parameters?

Currently, the tuner does not support additional parameters nor custom range. Such a feature is discussed in #1632.

> 4. Is early stopping implemented for the tuner? If so what's the early stopping logic?

The tuner itself does not have any logic to implement early stopping. It depends on the early stopping mechanism of the original LightGBM library.

**Comment by rohan-gt** [comment]
@toshihikoyanase wouldn't it be a useful feature to include early stopping for the tuner itself like within the steps of tuning each parameter using the overall train score vs. the CV score progression? So a user may give 1000 boosting rounds but because of early stopping, it can stop after 200 rounds

**Comment by toshihikoyanase** [comment]
@rohan-gt I'm sorry but I couldn't get the point. I think the following example uses early stopping with `LightGBMTunerCV`:

https://github.com/optuna/optuna/blob/dc032afdda0e0248bb70831ee47ca10aeca3bd28/examples/lightgbm_tuner_cv.py#L24-L26

When I execute it we can see that the tuner stopped, for example, after 300 rounds as follows:

```console
[100]   cv_agg's binary_logloss: 0.104948 + 0.0490855
[200]   cv_agg's binary_logloss: 0.0974624 + 0.0508658
[300]   cv_agg's binary_logloss: 0.113293 + 0.0718877
regularization_factors, val_score: 0.096504:   5%|#####1                                                                                                 | 1/20 
[00:00<00:03,  5.54it/s][I 2020-10-02 18:53:44,789] Trial 43 finished with value: 0.09710013822596493 and parameters: {'lambda_l1': 0.0021242176298051957, 'lambda_l2': 2.0198622863500958e-06}. Best is trial 
```

Is it insufficient for your use case?

**Comment by gaceladri** [comment]
I have another question related to LightGBMTunerCV. How could I get the study to use plot the results with the _visualization_ module? It is a feature already available or something that we should implement?

Thanks

**Comment by rohan-gt** [comment]
@toshihikoyanase What I was suggesting was two levels of early stopping. One to optimize `n_estimators` in LightGBM and the other to optimize `n_trials` in Optuna. So for if `n_trials=100`, you can calculate the cumulative min/max of the CV score of all the trials before it to perform early stopping
